pkg/infra/baseagent.go

```diff
@@ -0,0 +1,105 @@
+package infra
+
+import (
+	"container/list"
+	"errors"
+	"log"
+	"math"
+	"sync"
+
+	"github.com/SOMAS2021/SOMAS2021/pkg/messages"
+	"github.com/SOMAS2021/SOMAS2021/pkg/utils/abm"
+)
+
+type Base struct {
+	id        string
+	hp        int
+	floor     int
+	agentType int
+	inbox     *list.List
+	tower     *Tower
+	mx        sync.RWMutex
+}
+
+func NewBaseAgent(a *abm.ABM, aType int, agentHP int, agentFloor int, id string) (*Base, error) {
+	world := a.World()
+	if world == nil {
+		return nil, errors.New("agent needs a world defined to operate")
+	}
+	tower, ok := world.(*Tower)
+	if !ok {
+		return nil, errors.New("agent needs a tower world to operate")
+	}
+	return &Base{
+		id:        id,
+		hp:        agentHP,
+		floor:     agentFloor,
+		agentType: aType,
+		tower:     tower,
+		inbox:     list.New(),
+	}, nil
+}
+
+func (a *Base) Run() {
+	floor := a.floor
+	log.Printf("An agent cycle executed from base agent %d", floor)
+}
+
+func (a *Base) HP() int {
+	return a.hp
+}
+
+func (a *Base) Floor() int {
+	return a.floor
+}
+
+func (a *Base) ID() string {
+	return a.id
+}
+
+func (a *Base) IsAlive() bool {
+	_, found := a.tower.agents[a.id]
+	return found
+}
+
+func (a *Base) setFloor(newFloor int) {
+	a.floor = newFloor
+}
+
+func (a *Base) setHP(newHP int) {
+	a.hp = newHP
+}
+
+func (a *Base) updateHP(foodTaken float64) {
+	a.hp = int(math.Min(100, float64(a.hp)+foodTaken))
+}
+
+func (a *Base) TakeFood(amountOfFood float64) float64 {
+	if uint64(a.floor) == a.tower.currPlatFloor {
```

Comment:
Need to be careful with this cast, make sure that a.floor is not negative (which it shouldn't be)

Reply:
Fixed, thanks